mythril/laser/smt/solver.py

```diff
@@ -1,15 +1,16 @@
 """This module contains an abstract SMT representation of an SMT solver."""
 import z3
+from typing import Union, cast
 
 from mythril.laser.smt.expression import Expression
 
 
 class Solver:
     """An SMT solver object."""
 
-    def __init__(self):
+    def __init__(self) -> None:
         """"""
-        self.raw = z3.Solver()
+        self.raw = z3.Solver()  # type: Union[z3.Solver, z3.Optimize]
```

Comment:
Any reason why this should also have z3.Optimize?

Reply:
The issue is that the (mythril) Solver class is a super class of the (mythril) Optimize class, and they both define self.raw. Unfortunately, in z3 z3.Optimize is not a subclass of z3.Solver.

So the only way I could think of to unify this was with a union.

Something else that may work could be something along these lines:

T=TypeVar("T", bound=Union[z3.Solver, z3.Optimize])
class BaseSolver(Generic[T]):
    raw:  T

class Solver(BaseSolver[z3.Solver]):
    ...

class Optimize(BaseSolver[z3.Optimize]):
   ...

But then we would have to update the type of *all* places where Solver is expected to BaseSolver

Reply:
Yes, that would be much cleaner.

Reply:
So this is a bit tricky. As we have to by Python 3.5 compliant, I cannot use class variable type annotations like this:

```
class BaseSolver(Generic[T]):
      raw: T
```

Instead I need to add a dummy init and annotate:

```
class BaseSolver(Generic[T]):
    def __init__(self):
        self.raw = ?  # type: T
```

Now the issue becomes what the ? is. If i put None, then I must make the type Optional[T], which is incorrect and adds its own problems. An extremly ugly hack that works is:

```
class BaseSolver(Generic[T]):
    def __init__(self):
        self.raw = self.raw  # type: T
```
Ideas?

Reply:
Never mind, the answer is obvious..

Reply:
I fixed it in the next commit, adding a BaseSolver class from which both Solver and Optimize refer. Please note that I also chagned the signatures of the BaseSolver.{add,append} methods (perviously Solver.{add, append}) to make them tighter.